File: R/eqocean.R
cshift <- function(n, k) {
  ((1:n + k - 1) %% n + 1)
}

forward <- function(q, df, sgm, gma, tau) {
  n <- length(q)
  (q - 0.5 * gma * (q[cshift(n, 1)] - q[cshift(n, -1)]) +
    　 0.5 * gma^2 *(q[cshift(n, 1)] - 2 * q + q[cshift(n, -1)]) +
       tau * df) / (1 + sgm * tau)
}

adjoint <- function(p, h, eta, sgm, gma, epsh, tau) {
  n <- length(p)
  (p - 0.5 * gma * (p[cshift(n, -1)] - p[cshift(n, 1)]) +
       0.5 * gma^2 * (p[cshift(n, -1)] - 2 * p + p[cshift(n, 1)]) +
       epsh * (eta - h)
  ) / (1 + sgm * tau)
}

calc_h <- function(q0, q2, q4) q0 + q2 - q4

retrieve <- function(tdp, phi, fflg) {
  ifelse(fflg == 1, -tdp - phi, 0)
}

File: Python/eqocean.py
import numpy as np

def cshift(x, k):
    return np.roll(x, -k)

def forward(q, df, sgm, gma, tau): 
    return (q - 0.5 * gma * (cshift(q, 1) - cshift(q, -1)) +
            0.5 * gma ** 2 *(cshift(q, 1) - 2 * q + cshift(q, -1)) +
            tau * df) / (1 + sgm * tau)

def tangent(q, sgm, gma, tau):
    return (q - 0.5 * gma * (cshift(q, 1) - cshift(q, -1)) + 
            0.5 * gma ** 2 *(cshift(q, 1) - 2 * q + cshift(q, -1))) / (1 + sgm * tau)

def adjoint(p, dh, sgm, gma, eps, tau):
    return (p - 0.5 * gma * (cshift(p, -1) - cshift(p, 1)) +
            0.5 * gma ** 2 * (cshift(p, -1) - 2 * p + cshift(p, 1)) +
            eps * dh) / (1 + sgm * tau)

def calc_h(q0, q2, q4):
    return q0 + q2 - q4
